### PR TITLE
WDL 1.0 

### PR DESCRIPTION
**Pull request by ansingh7115**
fc-develop accompanying PR: https://github.com/broadinstitute/firecloud-develop/pull/1722

### **Context**
WDL 1.0 doesn't work in Terra because rawls and agora use an old womtool library to validate WDLs. Instead, we will be hitting the cromwell `/api/womtool/v1/describe` endpoint, which will return the results of validation. This means that we are passing the responsibility of validating WDLs to Cromwell and do will (hopefully) not need to worry when WDL versions are progressed.

### **What the Code did before**
The `MethodConfigResolver` object called the `CachingThreadPoolWDLParser` object to parse WDLs. The `CachingThreadPoolWDLParser` set up a Caffeine cache, used the `wdl-draft2` library to parse the wdl, and then saved it to the cache.

### **What the Code does now**
- Any evidence of the `wdl-draft2` library has been removed. 
- `CromwellSwaggerClient` was added to access the `/describe `endpoint
- The `WDLParser` trait was added, which is extended by the classes `CachingWDLParser` and `NonCachingWDLParser`, both of which call the `CromwellSwaggerClient` to parse WDLs. `CachingWDLParser` caches the results, `NonCachingWDLParser` does not.
- `MethodConfigResolver` is now an object instead of a class, so that an instance of a `WDLParser` can be passed to it.
- The `wdl-draft2` library objects and types (`InputDefinition`, `WomArrayType`, etc) have been replaced by the swagger generated classes (`WorkflowDescription`, `ToolInputParameter`, `ToolOutputParameter`)

### **Tests**
- A `MockCromwellSwaggerClient` was created. Tests in which WDLs were parsed require the creation of a `WorkflowDescription` response, which are added to the MockCromwellSwaggerClient before the test is run.
- Convenience methods for testing were added to the `MockCromwellSwaggerClient` companion object - they create instances of the Cromwell response : `makeWorkflowDescription`, `makeToolInputParameter`, etc.
- Cromwell returns WDL input and output names in the format "task.inputName" instead of "workflow.task.inputName", which required changing a lot of test data

### **Remaining Troubles**
- You'll see that a few tests are ignored. They are all failing because of a problem I have not yet been able to debug. All of the tests GET a method from Dockstore. In the tests, this is simulated by the adding expected responses to the RemoteMockServer. However, for some reason, this is returning a 404. 
- I have been creating FiaBs and then throwing different WDLs at rawls. Debugging an issue with WDLs with multiple inputs. 

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/MethodConfigResolver.scala [hunk 1]
```diff
@@ -45,30 +46,25 @@ object MethodConfigResolver {
     SubmissionValidationValue(attr, None, inputName)
   }
 
-  private def unpackResult(mcSequence: Iterable[AttributeValue], wfInput: InputDefinition): SubmissionValidationValue = wfInput.womType match {
-    case arrayType: WomArrayType => getArrayResult(wfInput.localName.value, mcSequence)
-    case WomOptionalType(_:WomArrayType) => getArrayResult(wfInput.localName.value, mcSequence) //send optional-arrays down the same codepath as arrays
-    case _ => getSingleResult(wfInput.localName.value, mcSequence, wfInput.optional)
+  private def unpackResult(mcSequence: Iterable[AttributeValue], wfInput: ToolInputParameter): SubmissionValidationValue = wfInput.getValueType.getTypeName match {
+      // TODO: Fix this match
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/MethodConfigResolver.scala]: does this still need doing?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/MethodConfigResolver.scala [hunk 2]
```diff
@@ -138,20 +135,33 @@ object MethodConfigResolver {
     }
   ) toString
 
-  def getMethodInputsOutputs(wdl: String): Try[MethodInputsOutputs] = parseWDL(wdl) map { parsedWdlWorkflow =>
-    val inputs = parsedWdlWorkflow.inputs map {
-      case (fqn: FullyQualifiedName, wfInput: InputDefinition) => model.MethodInput(fqn, wfInput.womType.stableName, wfInput.optional)
+  def getMethodInputsOutputs(userInfo: UserInfo, wdl: String)(implicit executionContext: ExecutionContext): Try[MethodInputsOutputs] = parseWDL(userInfo, wdl) map { workflowDescription =>
+    if (workflowDescription.getValid) {
+      val inputs = workflowDescription.getInputs.asScala.toList map { input =>
+        // TODO: getTypeName can return a type ("Int", "String") or "Optional" parse WorkflowDescription properly
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/MethodConfigResolver.scala]: does this still need doing?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/MethodConfigResolver.scala [hunk 3]
```diff
@@ -138,20 +135,33 @@ object MethodConfigResolver {
...
+        model.MethodInput(input.getName, input.getTypeDisplayName.replaceAll("\\n", ""), input.getOptional)
+      }
+      val outputs = workflowDescription.getOutputs.asScala.toList map { output =>
+        model.MethodOutput(output.getName, output.getTypeDisplayName.replace("\\n", ""))
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/MethodConfigResolver.scala:147]: why do we need to replace \n?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/CachingWDLParser.scala [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  implicit val customisedCaffeineCache: Cache[Try[WorkflowDescription]] = CaffeineCache(underlyingCaffeineCache)
+
+
+  override def parse(userInfo: UserInfo, wdl: String)(implicit executionContext: ExecutionContext): Try[WorkflowDescription] = {
+    val tick = System.currentTimeMillis()
+    val key = generateCacheKey(wdl)
+
+    get(key) match {
+      case Some(parseResult) =>
+        val tock = System.currentTimeMillis() - tick
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/CachingWDLParser.scala:40]: do you need tick and tock?

**Review by dvoet — APPROVED** [review]
a couple small comments

File: core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala [hunk 5]
```diff
@@ -609,7 +611,7 @@ class SubmissionSpec(_system: ActorSystem) extends TestKit(_system) with FlatSpe
     }
 
     assert {
-      rqComplete.errorReport.message.contains("Invalid outputs: three_step.cgrep.count")
+      rqComplete.errorReport.message.contains("Invalid outputs: cgrep.count")
     }
```
Comment (kcibul) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala]: Just curious -- why the changes for three_step -> gcrep ?

File: project/Dependencies.scala [hunk 6]
```diff
@@ -72,15 +74,7 @@ object Dependencies {
   val mockserverNetty: ModuleID = "org.mock-server"               % "mockserver-netty"      % "3.9.2" % "test"
   val ficus: ModuleID =           "com.iheart"                    %% "ficus"                % "1.4.0"
   val scalaCache: ModuleID =      "com.github.cb372"              %% "scalacache-caffeine"  % "0.24.2"
-
-  val cromwellWdl: ModuleID = ("org.broadinstitute" %% "wdl-draft2" % cromwellVersion
```
Comment (kcibul) [inline project/Dependencies.scala:76]: yaay!

**Review by kcibul — APPROVED** [review]
lgtm -- assuming that you resolve the failing/ignored tests you mentioned in the PR comment (maybe that was resolved as described in the URL decoding comments in the code?)

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/CachingWDLParser.scala [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        parseResult
+      case None => parseAndCache(userInfo, wdl, key)
+    }
+  }
+
+  private def parseAndCache(userInfo: UserInfo, wdl: String, key: String)(implicit executionContext: ExecutionContext): Try[WorkflowDescription] = {
+    val parseResult = inContextParse(userInfo, wdl)
+    val timeToLive = parseResult match {
+      case Success(_) => Some(wdlParsingConfig.cacheTTLSuccessSeconds)
+      case Failure(ex) => Some(wdlParsingConfig.cacheTTLFailureSeconds)
```
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/CachingWDLParser.scala]: I think the failure case should log the exception and any other helpful debug info, especially with womtool being pretty new.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/CromwellSwaggerClient.scala [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    val wfdescriptionInputs  = wfdescription.getInputs.asScala.toList.map { input =>
+      val rawInput = input.getName
+      input.name(wfdescription.getName + "." + rawInput)
+    }
+    val wfdescriptionOutputs = wfdescription.getOutputs.asScala.toList map { output =>
+      val rawOutput = output.getName
+      output.name(wfdescription.getName + "." + rawOutput)
+    }
+    wfdescription.inputs(wfdescriptionInputs.asJava).outputs(wfdescriptionOutputs.asJava)
+  }
```
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/CromwellSwaggerClient.scala:27]: I *think* that the Scala compiler will optimize for this, but consider calling `wfdescription.getName` once, outside of the loops. I realize this is probably a micro-optimization:
```
    val wfname:String = wfdescription.getName
    val wfdescriptionInputs = wfdescription.getInputs.asScala.map { input =>
      input.name(s"$wfname.${input.getName}")
    }
    val wfdescriptionOutputs = wfdescription.getOutputs.asScala map { output =>
      output.name(s"$wfname.${output.getName}")
    }
```

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/CromwellSwaggerClient.scala [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+  private def getCromwellWomtoolApi(accessToken: String): WomtoolApi = {
+    val apiClient = new ApiClient()
+    apiClient.setAccessToken(accessToken)
+    apiClient.setBasePath(cromwellBasePath)
+    new WomtoolApi(apiClient)
+  }
+
+
+  // CHANGING THIS MOMENTARILY TO TEST A BUG
```
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/CromwellSwaggerClient.scala]: is this comment still valid?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/CachingWDLParser.scala [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    * this method exists as an abstraction, making it easy to change what we use as a cache key in case
+    * we want to reduce large wdl payloads to something smaller. Current implementation returns the wdl
+    * unchanged to ensure correctness of cache lookups.
+    *
+    * @param wdl
+    * @return
+    */
+  private def generateCacheKey(wdl: String): String = {
+    wdl
+  }
```
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/CachingWDLParser.scala:99]: I would love to see this report quantitative metrics somewhere - potentially, custom metrics in New Relic:
* cache hits vs. cache misses
* time spent in `parse` for cache hits
* time spent in `parse` for cache misses
* time spent in `cromwellSwaggerClient.describe` (should be very close to cache-miss time)
* number of errors? not sure what else.

in the now-deleted `CachingThreadPoolWDLParser`, the `logger.info(s"<parseWDL-cache> ...` log entries went to Kibana, where a dashboard exists to show cache hits. vs misses.

I'm fine if this is a follow-on PR, but I think we'll want these metrics very close to when we launch this to prod so we can understand how well WaaS is working.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/CachingWDLParser.scala [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    }
+    put(key)(parseResult, ttl = timeToLive)
+    parseResult
+  }
+
+
+
+  private def inContextParse(userInfo: UserInfo, wdl: String)(implicit executionContext: ExecutionContext): Try[WorkflowDescription] = {
+   Try { cromwellSwaggerClient.describe(userInfo, wdl) }
+  }
```
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/CachingWDLParser.scala:86]: should this have any retry logic?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/NonCachingWDLParser.scala [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+import scala.util.Try
+
+class NonCachingWDLParser(wdlParsingConfig: WDLParserConfig, cromwellSwaggerClient: CromwellSwaggerClient) extends WDLParser {
+
+  def parse(userInfo: UserInfo, wdl: String)(implicit executionContext: ExecutionContext): Try[WorkflowDescription] = {
+    Try { cromwellSwaggerClient.describe(userInfo, wdl) }
+  }
+
+
+}
```
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/NonCachingWDLParser.scala:20]: could/should this class disappear, and instead have `CachingWDLParser` accept a constructor argument of `useCache: Boolean`? (and potentially rename "CachingWDLParser")

If you end up adding custom metrics reporting, retry logic, and/or logging, you'd have to copy-paste it to both classes.

File: project/Dependencies.scala [hunk 13]
```diff
@@ -29,6 +29,8 @@ object Dependencies {
   val akkaTestKit: ModuleID =       "com.typesafe.akka"   %%  "akka-testkit"         % akkaV     % "test"
   val akkaHttpTestKit: ModuleID =   "com.typesafe.akka"   %%  "akka-http-testkit"    % akkaHttpV % "test"
 
+  val cromwellClient: ModuleID =    "org.broadinstitute.cromwell" %% "cromwell-client" % "0.1-15a30475d-SNAP"
+
```
Comment (davidangb) [inline project/Dependencies.scala:33]: is there a non-SNAP version to use?

**Review by davidangb — COMMENTED** [review]
comments inline ...

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/NonCachingWDLParser.scala (hunk 12 above)
Comment (ansingh7115) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/NonCachingWDLParser.scala:20]: I think there are trade-offs to having 2 classes for caching and non-caching or 1 class with a useCache argument.

If we have 2 classes, we keep the code for the caching and non-caching pathways separate, which I thing would help for debugging. It also means we're not evaluating a `useCache` argument for each call. We also avoid more code complexity around whether or not we should instantiate the cache. 

If we have both caching and noncaching in the same class, we have fewer files and metrics, retries and logging are in the same place. 

However, I think I can move the retrying to the CromwellSwaggerClient and remove it from here. And I think, since the noncaching parser is just calling the client, it would have minimal logging and metrics anyway.

So, in conclusion, my instinct is to keep it as 2 different classes.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/MethodConfigResolver.scala (hunk 3 above)
Comment (ansingh7115) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/MethodConfigResolver.scala:147]: talked about face to face - certain `TypeDisplayName`s for Composite Types have new lines which we don't want to return

File: project/Dependencies.scala (hunk 13 above)
Comment (ansingh7115) [reply project/Dependencies.scala:33]: no - cromwell doesn't currently automatically build and push to artifactory yet

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/CachingWDLParser.scala (hunk 10 above)
Comment (ansingh7115) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/CachingWDLParser.scala:99]: I'll have to learn more about how new relic works to do what you're talking about, I can add better logging for now and then create a ticket for the new relic stuff as follow-on work!

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/NonCachingWDLParser.scala (hunk 12 above)
Comment (davidangb) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/NonCachingWDLParser.scala:20]: works for me - I don't feel super strong about this

File: project/Dependencies.scala (hunk 13 above)
Comment (davidangb) [reply project/Dependencies.scala:33]: gotcha, thanks

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/CachingWDLParser.scala (hunk 10 above)
Comment (ansingh7115) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/wdlparsing/CachingWDLParser.scala:99]: Adding back the log entries!

File: core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala (hunk 5 above)
Comment (ansingh7115) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala]: changed it back - it was originally because cromwell was returning inputs and outputs without the workflow name appended

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/MockCromwellSwaggerClient.scala [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+  val workflowDescriptions: mutable.Map[String, WorkflowDescription] =  new TrieMap()
+
+  override def describe(userInfo: UserInfo, wdl: String): Try[WorkflowDescription] = {
+    Try { MockCromwellSwaggerClient.returnCopy(workflowDescriptions(wdl)) }
+  }
+
+}
+
+object MockCromwellSwaggerClient {
+  def returnCopy(wfDescription: WorkflowDescription): WorkflowDescription = {
```
Comment (ansingh7115) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/MockCromwellSwaggerClient.scala:23]: Added this because `workflowDescription` above was passing around a references to objects, which meant that tests were changing the WorkflowDescriptions in the TrieMap

File: core/src/test/scala/org/broadinstitute/dsde/rawls/mock/RemoteServicesMockServer.scala [hunk 15]
```diff
@@ -255,8 +175,7 @@ class RemoteServicesMockServer(port:Int) extends RawlsTestUtils {
     mockServer.when(
       request()
         .withMethod("GET")
-        // Apparently the mock server url-decodes paths before comparing
-        .withPath("/ga4gh/v1/tools/#workflow/dockstore-method-path/versions/dockstore-method-version/WDL/descriptor")
+        .withPath("/ga4gh/v1/tools/%23workflow%2Fdockstore-method-path/versions/dockstore-method-version/WDL/descriptor")
```
Comment (davidangb) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/mock/RemoteServicesMockServer.scala:178]: why did this change? This feels like a functional difference.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/mock/RemoteServicesMockServer.scala [hunk 16]
```diff
@@ -306,13 +210,10 @@ class RemoteServicesMockServer(port:Int) extends RawlsTestUtils {
...
     )
 
-    val noInputMethodDockstoreResponse =
-      s"""{"type":"WDL","descriptor":"${noInputWdl.replace("t1", "t1_dockstore").replace("\"", "\\\"").replace("\n","\\n")}","url":"bogus"}"""
-
     mockServer.when(
       request()
         .withMethod("GET")
-        .withPath("/ga4gh/v1/tools/#workflow/dockstore-no-input-path/versions/dockstore-no-input-version/WDL/descriptor")
+        .withPath("/ga4gh/v1/tools/%23workflow%2Fdockstore-no-input-path/versions/dockstore-no-input-version/WDL/descriptor")
```
Comment (davidangb) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/mock/RemoteServicesMockServer.scala:216]: same  here - why did this change?

File: core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/MethodConfigApiServiceSpec.scala [hunk 17]
```diff
@@ -69,11 +69,11 @@ class MethodConfigApiServiceSpec extends ApiServiceSpec {
   }
 
   "MethodConfigApi" should "return 201 on create method configuration in Agora" in withTestDataApiServices { services =>
-    testCreateMethodConfiguration(AgoraMethod("dsde", "three_step", 1), "three_step", services)
+    testCreateMethodConfiguration(AgoraMethod("dsde", "three_step", 1), "three_step_dockstore", services)
   }
 
   it should "return 201 on create method configuration in Dockstore" in withTestDataApiServices { services =>
-    testCreateMethodConfiguration(DockstoreMethod("dockstore-method-path", "dockstore-method-version"), "three_step_dockstore", services)
+    testCreateMethodConfiguration(AgoraMethod("dsde", "three_step", 1), "three_step", services)
```
Comment (davidangb) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/MethodConfigApiServiceSpec.scala]: this should really be testing a `DockstoreMethod`. If for some reason that's not possible (why?),  just delete the test, since it duplicates the one above it

File: core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/MethodConfigApiServiceSpec.scala [hunk 18]
```diff
@@ -69,11 +69,11 @@ class MethodConfigApiServiceSpec extends ApiServiceSpec {
   }
 
   "MethodConfigApi" should "return 201 on create method configuration in Agora" in withTestDataApiServices { services =>
-    testCreateMethodConfiguration(AgoraMethod("dsde", "three_step", 1), "three_step", services)
+    testCreateMethodConfiguration(AgoraMethod("dsde", "three_step", 1), "three_step_dockstore", services)
```
Comment (davidangb) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/MethodConfigApiServiceSpec.scala]: why is this renamed "dockstore"?

**Review by davidangb — COMMENTED** [review]
a few comments in test code, looking very close.

I see that a lot of the input/output names have changed in the test code, mostly due to the workflow name being prepended. Does anything change for the end user?

File: core/src/test/scala/org/broadinstitute/dsde/rawls/mock/RemoteServicesMockServer.scala (hunk 15 above)
Comment (ansingh7115) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/mock/RemoteServicesMockServer.scala:178]: We think there might have been some dependency issue  with regards to taking the womtool library out - I talked to Doug about it and he said not to worry about it

File: core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/MethodConfigApiServiceSpec.scala (hunk 17 above)
Comment (ansingh7115) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/MethodConfigApiServiceSpec.scala]: Oh shoot, thanks for catching that

File: core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/MethodConfigApiServiceSpec.scala (hunk 18 above)
Comment (ansingh7115) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/MethodConfigApiServiceSpec.scala]: Yeah, I think I messed up when I was reverting some changes - going to go through the tests one more time to make sure it's right, thank you!

**Comment by ansingh7115** [comment]
>I see that a lot of the input/output names have changed in the test code, mostly due to the workflow name being prepended. Does anything change for the end user?

Cromwell's responses don't come with the workflow prepended, so I'm added them on  the WDLParser level (which gets the unprepended response from CromwellSwaggerClient). So the tests shouldn't need to change that much. Originally, I had thought maybe we could make rawls just not expect the workspace name prepended (which is why I had originally changed so many tests) but then decided that change was too big and bulky when people seem to real want WDL 1.0 now. So I just prepended what we get from Cromwell

**Comment by ansingh7115** [comment]
jenkins retest

**Comment by davidangb** [comment]
@ansingh7115 are you saying that WaaS returns different values (unprepended) than the wdl-draft2 library parser did? If so, then I (finally) understand why you had to change things internal to rawls.

**Comment by ansingh7115** [comment]
Yes :) I'm sorry that wasn't clear!

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/MockExecutionServiceDAO.scala [hunk 19]
```diff
@@ -25,7 +24,7 @@ class MockExecutionServiceDAO(timeout:Boolean = false, val identifier:String = "
     this.collection = workflowCollection
 
 
-    val inputPattern = """\{"three_step.cgrep.pattern":"(sample[0-9])"\}""".r
+    val inputPattern = """\{"three.step.cgrep.pattern":"(sample[0-9])"\}""".r
```
Comment (davidangb) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/MockExecutionServiceDAO.scala]: `three.step`, not `three_step`?

**Review by davidangb — APPROVED** [review]
:+1: once tests pass

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/MockExecutionServiceDAO.scala (hunk 19 above)
Comment (ansingh7115) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/MockExecutionServiceDAO.scala]: Oh my god, I'm insane
